### PR TITLE
Add the rest of the test services

### PR DESCRIPTION
**Pull request by jackkleeman**
Except the workflow one as we don't have that yet

Also fix a few bugs to make them pass:
1. Not handling partial state flag correctly
2. Not handling failures on sleeps (currently this happens due to cancellation). I have opted to make sleep return an error to implement this, but I expect that people will pretty regularly ignore that error.

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -19,8 +19,8 @@ type Context interface {
 	// This source is not safe for use inside .Run()
 	Rand() *rand.Rand
 
-	// Sleep for the duration d
-	Sleep(d time.Duration)
+	// Sleep for the duration d. Can return a terminal error in the case where the invocation was cancelled mid-sleep.
+	Sleep(d time.Duration) error
```
Comment (slinkydeveloper) [inline context.go:23]: I now understand the point you had about cancellations in the DM.

Then yes i wonder if we could plug the cancellations to the go context cancellation somehow... (let's not do it in this pr though)

File: test-services/exclusions.yaml
Comment (slinkydeveloper) [inline test-services/exclusions.yaml:1]: :heart_eyes:

File: test-services/nondeterministic.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	doLeftAction := func(ctx restate.ObjectContext) bool {
+		countKey := ctx.Key()
+		invocationCountsMtx.Lock()
+		defer invocationCountsMtx.Unlock()
+
+		invocationCounts[countKey] += 1
+		return invocationCounts[countKey]%2 == 1
+	}
+	incrementCounter := func(ctx restate.ObjectContext) error {
+		return ctx.Object("Counter", "abc", "add").Send(int64(1), 0)
```
Comment (slinkydeveloper) [inline test-services/nondeterministic.go:26]: Only for this increment, the Counter key here must be the `ctx.key()`, that's what the test runner will check for.

File: test-services/testutils.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+							Type:  "result",
+							Value: &result,
+						}, nil
+					}
+
+					timeout := ctx.After(time.Duration(*req.AwaitTimeout) * time.Millisecond)
+					selector := ctx.Select(timeout, awakeable)
+					switch selector.Select() {
+					case timeout:
+						return CreateAwakeableAndAwaitItResponse{Type: "timeout"}, nil
```
Comment (slinkydeveloper) [inline test-services/testutils.go:62]: (API comment, not to fix in this PR)
Perhaps we could have something like ctx.WithTimeout for a future with a timeout?

**Review by slinkydeveloper — COMMENTED** [review]
Just saw one minor detail in the nondeterministic test that is wrong, everything else looks good! Thanks for doing this!

File: test-services/testutils.go (hunk 3 above)
Comment (jackkleeman) [reply test-services/testutils.go:62]: or possibly as a method on the responsefuture/awakeable

**Pull request merged**